src/Disciplina/DB/DSL/Types.hs

```diff
@@ -0,0 +1,54 @@
+{-# LANGUAGE TypeOperators #-}
+
+module Disciplina.DB.DSL.Types
+        ( WHERE (..)
+        , TxIdEq (..)
+        , TxGrade (..)
+        , NodeWithProp (..)
+        , TxsFilterExpr (..)
+        , ObjHashEq (..)
+        , QueryTx (..)
+        , QueryTxs (..)
+        , QueryObj (..)
+        )
+        where
+
+import Universum
+import Disciplina.Educator.Txs (PrivateTxId)
+import Disciplina.Crypto (Hash)
+import qualified Disciplina.Core as Core (Grade, SubjectId)
+
+data WHERE = WHERE
+
+data TxIdEq = TxIdEq PrivateTxId
+
+data TxGrade = TxGrade
+
+data NodeWithProp = SubjectNode Core.SubjectId
+                  | TxSubjectNode
+
+-- | TODO, what is the correct type for TxSubjectIdEq?
+data TxsFilterExpr = TxSubjectIdEq Int
+                   | TxGradeEq Core.Grade
+                   | (:>=) TxGrade Core.Grade
+                   | (:&) TxsFilterExpr TxsFilterExpr
+                   | (:||) TxsFilterExpr TxsFilterExpr
+                   | (:~>) NodeWithProp NodeWithProp
```

Comment:
It actually makes more sense to have something like `TxSubjectIsDescendantOf Core.SubjectId` instead here. With this constructor it's possible to have expression like `SubjectNode 3 :~> SubjectNode 5`, which doesn't represent any predicate on a transaction.